### PR TITLE
feat: implement first-line-regex 

### PR DESCRIPTION
**Pull request by sigmaSd**
fix https://github.com/tree-sitter/tree-sitter/issues/2478

I added a PR for helix for the same thing https://github.com/helix-editor/helix/pull/7853

But there I went with a Vec<String> instead of a String, the example is strace output it can start with `["execve", ,"^[0-9:.]* *execve", "^__libc_start_main"]`

Obviously this is not a structured language by any mean (though I wrote a wip grammar for it) so it might be an invalid argument

### CODE REVIEW

File: cli/loader/src/lib.rs
Comment (amaanq) [inline cli/loader/src/lib.rs:109]: It'd be better to do away with this and perhaps just use the string version of the regex for the hashmap entry, and then use Regex::from_str to convert back when using `find`

**Comment by amaanq** [comment]
Also a test or two could be useful if you don't mind adding that

File: cli/loader/src/lib.rs
Comment (amaanq) [reply cli/loader/src/lib.rs:109]: Here's what I was thinking:

```patch
diff --git a/cli/loader/src/lib.rs b/cli/loader/src/lib.rs
index 0f9a33fe..88c47986 100644
--- a/cli/loader/src/lib.rs
+++ b/cli/loader/src/lib.rs
@@ -8,6 +8,7 @@ use std::io::BufReader;
 use std::ops::Range;
 use std::path::{Path, PathBuf};
 use std::process::Command;
+use std::str::FromStr;
 use std::sync::Mutex;
 use std::time::SystemTime;
 use std::{env, fs, mem};
@@ -93,26 +94,12 @@ pub struct LanguageConfiguration<'a> {
     use_all_highlight_names: bool,
 }
 
-#[derive(Debug)]
-struct Regex2(Regex);
-impl Eq for Regex2 {}
-impl PartialEq for Regex2 {
-    fn eq(&self, other: &Self) -> bool {
-        self.0.as_str() == other.0.as_str()
-    }
-}
-impl std::hash::Hash for Regex2 {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.0.as_str().hash(state)
-    }
-}
-
 pub struct Loader {
     parser_lib_path: PathBuf,
     languages_by_id: Vec<(PathBuf, OnceCell<Language>)>,
     language_configurations: Vec<LanguageConfiguration<'static>>,
     language_configuration_ids_by_file_type: HashMap<String, Vec<usize>>,
-    language_configuration_ids_by_first_line_regex: HashMap<Regex2, Vec<usize>>,
+    language_configuration_ids_by_first_line_regex: HashMap<String, Vec<usize>>,
     highlight_names: Box<Mutex<Vec<String>>>,
     use_all_highlight_names: bool,
     debug_build: bool,
@@ -238,14 +225,22 @@ impl Loader {
                     })
             })
             .or_else(|| {
-                let Ok(file) = fs::File::open(path) else {return None};
+                let Ok(file) = fs::File::open(path) else {
+                    return None;
+                };
                 let reader = BufReader::new(file);
-                let Some(Ok(first_line)) = std::io::BufRead::lines(reader).next() else {return None};
+                let Some(Ok(first_line)) = std::io::BufRead::lines(reader).next() else {
+                    return None;
+                };
 
                 self.language_configuration_ids_by_first_line_regex
                     .iter()
-                    .find(|(regex,_)| dbg!(&regex.0).is_match(&first_line))
-                    .map(|(_,ids)| ids)
+                    .find(|(regex, _)| {
+                        Regex::from_str(regex)
+                            .and_then(|r| Ok(r.is_match(&first_line)))
+                            .unwrap_or(false)
+                    })
+                    .map(|(_, ids)| ids)
             });
 
         if let Some(configuration_ids) = configuration_ids {
@@ -616,7 +611,7 @@ impl Loader {
                     }
                     for first_line_regex in &configuration.first_line_regex {
                         self.language_configuration_ids_by_first_line_regex
-                            .entry(Regex2(first_line_regex.clone()))
+                            .entry(first_line_regex.to_string())
                             .or_insert(Vec::new())
                             .push(self.language_configurations.len());
                     }

```

Comment (sigmaSd) [reply cli/loader/src/lib.rs:109]: Should I just change https://github.com/tree-sitter/tree-sitter/blob/e854fc4731c20d56a6212569c36823a1aa84b166/cli/loader/src/lib.rs#L81 to String

Instead of the current Regex -> String -> Regex ?

Comment (sigmaSd) [reply cli/loader/src/lib.rs:109]: we loose the regex validation though

**Comment by sigmaSd** [comment]
todo
- [ ] tests

File: cli/loader/src/lib.rs
Comment (amaanq) [reply cli/loader/src/lib.rs]: Up to you, but I don't really like the whole abstraction of a struct on top of regex..

**Comment by sigmaSd** [comment]
I applied your suggestion and added a tes

File: cli/src/tests/helpers/dirs.rs
Comment (sigmaSd) [inline cli/src/tests/helpers/dirs.rs:6]: is this ok ?

Comment (amaanq) [reply cli/src/tests/helpers/dirs.rs:6]: maybe `pub(crate)` to avoid accidental exports in the future?

Comment (amaanq) [reply cli/src/tests/helpers/dirs.rs:6]: Instead - maybe we can use python and set the regex to check for `#!` and then python somewhere after.

Comment (amaanq) [reply cli/src/tests/helpers/dirs.rs]: I've pushed a better change for this, if you're good w/ that do you mind squashing the commits down to 1? Or I can if you're wary about doing that.

Comment (sigmaSd) [reply cli/src/tests/helpers/dirs.rs]: Yeah it's fine I leave that to you  I'm used to commiting random stuff and the maintainers squash and merge

File: cli/loader/src/lib.rs
Comment (ahlinc) [inline cli/loader/src/lib.rs:82]: On this place the `Regex` type should be stored.

File: cli/loader/src/lib.rs
Comment (ahlinc) [inline cli/loader/src/lib.rs:239]: See above comment, it's not a good idea to instantiate Regex instances here every time.

File: cli/loader/src/lib.rs
Comment (ahlinc) [inline cli/loader/src/lib.rs:614]: Look on above loop and how a `Regex` converted to a `String` there.

File: cli/loader/src/lib.rs
Comment (ahlinc) [inline cli/loader/src/lib.rs:243]: I think all this implementation should be improved by using a `RegexSet`: https://docs.rs/regex/1.9.3/regex/struct.RegexSet.html

**Review by ahlinc — CHANGES_REQUESTED** [review]

File: cli/loader/src/lib.rs
Comment (amaanq) [reply cli/loader/src/lib.rs:243]: Does regexset implement Hash? That's the main issue

Comment (ahlinc) [reply cli/loader/src/lib.rs:243]: It's an additional note about implementation efficiency, it can be ignored cause requires more redesign, I can do it by myself later.

**Comment by amaanq** [comment]
thanks, sorry for taking so long to get to this.

**Pull request merged**
